cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/ProgramGenerationStage.java

```diff
@@ -61,16 +65,24 @@ public void process(final ApplicationDeployable input) throws Exception {
     final ApplicationSpecification appSpec = input.getSpecification();
 
     // Make sure the namespace directory exists
-    Id.Namespace namespaceId = input.getApplicationId().getParent().toId();
-    Location namespacedLocation = namespacedLocationFactory.get(namespaceId);
-    // Note: deployApplication/deployAdapters have already checked for namespaceDir existence, so not checking again
-    // Make sure we have a directory to store the original artifact.
-    final Location appFabricDir = namespacedLocation.append(configuration.get(Constants.AppFabric.OUTPUT_DIR));
+    final Id.Namespace namespaceId = input.getApplicationId().getParent().toId();
+
+    impersonator.doAs(namespaceId.toEntityId(), new Callable<Void>() {
```

Comment:
We probably don't need this dir creation and check at all. Can you check if the artifact repo is actually writing anything to this directory? If not, I think we can remove this code block entirely. Still, if it does, then it should be the artifact repo who do the dir creation.